### ISSUE TITLE
Bug in Fl_Tabs::tab_positions()

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#718]
[Edited by Albrecht for clarity]
@pvrose wrote on April 11, 2023  in fltk.general in thread "[Possible bug in FlTabs::tabs_position()](https://groups.google.com/g/fltkgeneral/c/QAflXPOifq8/m/XWcKjDhLBgAJ)":
[End of Edit]

I have inherited Fl_Tabs for my own tabs widget. When I display it I occasionally get one of the tabs not displayed. I tried to understand the algorithm in tabs_position() that adjusts the tab widths if they are larger than the Fl_Tabs width but failed. I overrode tabs_position() to implement a different algorithm. 

Keep the selected tab width as calculated and then shrink the other tab widths by the same ratio.

I replaced the correction code with the following.

	// Change the algorithm if they are too big
	// Reduce all deselected widths by the apportioned oversize
	// Use fixed point arithmetic (x2^10) to reduce rounding error
	int delta = 1024 * (r - tab_width[selected]) / (tab_pos[i] - tab_width[selected]);
	for (i = 0; i < nc; i++) {
		if (i != selected) {
			tab_width[i] = tab_width[i] * delta / 1024;
		}
		tab_pos[i + 1] = tab_pos[i] + tab_width[i] + BORDER;
	}
	// If we haven't used the full width, increase the size of the last tab to compensate
	if (tab_pos[i] < r) {
		tab_width[i - 1] = r - tab_pos[i];
		tab_pos[i] = r;
	}
	return selected;

I am using a fairly old version of fLTK 1.4. It doesn't yet have the change that introduced the when(FL_WHEN_CLOSED) feature for Fl_Tabs.

**Comment by pvrose** [comment]
The problem is stiill there in the download I took today: "C:\Users\pvros\Downloads\fltk-1.4.x-20230407-6f0e9b65.tar.gz".

Please see the attached file:

[ttabs.cpp.txt](https://github.com/fltk/fltk/files/11213392/ttabs.cpp.txt)

If you select tab Four, tab Three is not visible.

Phil

**Comment by Albrecht-S** [comment]
I can confirm the bug with the given test program in [this comment above](https://github.com/fltk/fltk/issues/718#issuecomment-1505518272) and current FLTK (Git commit ad5e49464b4ba5972d2dbce762b02009755338d6) under Linux (Wayland). It is as described in the mentioned comment: tab "Three ..." disappears when tab "Four ..." is activated.

I set labels "bug" and "Prio 3 - Moderate" because an essential information (an existing tab) is missing.

I did not test the given patch though. I believe @MatthiasWM would know best what to do. Matthias?

**Comment by MatthiasWM** [comment]
Thanks for updating comments in the code, Albrecht. I got stuck when I tried to fix it and decided for a rewrite, but got sidetracked. I will try to make some time tomorrow.

**Comment by Albrecht-S** [comment]
> Thanks for updating comments in the code, Albrecht.

Yeah, when I was looking at the code I found that I needed better docs and tried to add docs for the related protected methods - at least as far as I could find out quickly what they were doing. Other protected methods (some are old, some were added recently) still need docs but I lost steam.

I believe we should really document the interface of such protected methods just in case someone wants (or needs) to derive a subclass. Undocumented protected methods might otherwise lead to chaos in user code and would be hard to fix or change later.

**Comment by MatthiasWM** [comment]
Ok, so I rewrote the overlapping algorithm. This one is tested with quite a bunch of extreme cases and can still keep all tabs selectable, even if two labels are too huge to even fit the width of Fl_Tabs. Hope this fixes it for good, because the whole thing is somewhat tricky to make user friendly. Ideally, OVERFLOW_PULLDOWN or OVERFLOW_DRAG should be set if there is a high number of tabs to be expected

**Comment by Albrecht-S** [comment]
Reopened. Unfortunately the latest update creates lots of artifacts:
- above visible tabs
- between visible tabs

See screenshot from the unchanged above test program after clicking several tabs (produced with scaling factor 200%):
![image](https://github.com/fltk/fltk/assets/26235500/60cc18b0-e98d-4d91-ab16-098f46bc4c5b)

Looks like damaging the widget (`damage_tabs()`?) doesn't work as intended.

Additional suggestion from testing (see screenshot): the label of the currently inactive tab "Four ..." should probably be left aligned. It would be more intuitive to see the first part of the label than anything else.

Note: ISTR that there is a tab label alignment flag, but I just verified that the test program doesn't set an alignment, hence this seems to be the default.

**Comment by MatthiasWM** [comment]
Oh, ok. At least the tab positions and sizes are as intended. No idea how I trampled on the clipping code, but I am sure I can find the issue. As for the alignment, the method to get and set alignment is only useful if you want and image above text, as I did in the FLUID Setting dialog. As the "Four" tab in this case is visually "under" the "Three" tab, I feel that it would look wrong if I aligned the label to the right edge of the previous tab. Granted, it would help readability.

**Comment by MatthiasWM** [comment]
LOL, ok. So in the test the children of Fl_Tabs are Fl_Buttons. If they are Fl_Groups, as they usually are, everything is 99% fine. Looking at `Fl_Tabs::draw()`, I can not immediately see why that is the case. So my changes did not bring the drawing error, but it probably was always there for Fl_Buttons.

It doesn't matter. `Fl_Tabs::draw()` is a mess. I will try to fix the (ab)use of damage flags ASAP and make sure the tabs section get a full redraw.

**Comment by Albrecht-S** [comment]
> As the "Four" tab in this case is visually "under" the "Three" tab, I feel that it would look wrong if I aligned the label to the right edge of the previous tab ...

Hmm, I didn't notice that the tabs are drawn in full size but "below and above" each other. Is this really what you intended?

I would have thought that the tab *sizes* were adjusted accordingly (as they were in the "original" version IIRC). If the positions **and** sizes were modified and the label would be left aligned and clipped to the tab size, then it would work as I assumed. Anyway, looking forward to your upcoming improvements. Many thanks!

**Comment by Albrecht-S** [comment]
> I would have thought that the tab sizes were adjusted accordingly (as they were in the "original" version IIRC).

Well, I didn't remember this correctly. I just checked with test/tabs from 1.3 (git):

![image](https://github.com/fltk/fltk/assets/26235500/4174e779-4b30-487b-b030-cddb33d12fac)

As you can (not!) see: Tab `tab4` is hidden between `tab3` and `print`.

Current FLTK 1.4 (git) is already a big improvement because it does no longer hide existing tabs:

![image](https://github.com/fltk/fltk/assets/26235500/af212615-3144-4a8b-b46e-9f18dec7d013)

I believe now:
- the clipping issue should be resolved
- better label alignment would be "nice to have" but it's not important (not a regression).

**Comment by MatthiasWM** [comment]
Yes, it always draws the entire tab, clipped by what is actually visible. I am in the process of cleaning up the ::draw method to  ensure that the tabs area is always completely redrawn. We tried to only partially update, but that draws the wrong background, or none, as we see in the example code, when the children are not derived from Fl_Group.

This is a pretty good chance to allow left and right side tabs as well, if that is a worthy feature, but as always, it adds complexity.

https://media.istockphoto.com/id/173620982/photo/address-book-emails.jpg?s=612x612&w=0&k=20&c=evNLIMG4iRDMYz0GGvbuAQlXUkBi_mWkBuIxKxfYYNw=

**Comment by MatthiasWM** [comment]
OK. I rewrote the `::draw()` method to be more readable, more efficient, and to no longer leave breadcrumbs behind without flickering too much on single buffer windows (anyone still using those?). Tested for all Schemes and for many color combinations, tested for top and bottom tabs.

Please close when happy.

**Comment by Albrecht-S** [comment]
Thanks, I'm "almost happy" ;-) but there's still a tiny issue, obviously in the overflow code. See screenshot:
![tabs](https://github.com/fltk/fltk/assets/26235500/d4329bfa-e50a-428b-ae49-29c4a31de837)

This happens **if the second tab is wider than the first tab**, the first tab is selected, and the Fl_Tabs width is too small so the compression mode kicks in. The tab position of the 2nd tab is obviously calculated too far left (i.e. one or two pixels left of the first tab).

Test case: use `test/tabs.fl` with the following diff, build it, and shrink the window size while the first tab is selected:
```diff
diff --git a/test/tabs.fl b/test/tabs.fl
index 56aec468b..28733f604 100644
--- a/test/tabs.fl
+++ b/test/tabs.fl
@@ -42,7 +42,7 @@ Function {} {open
         }
       }
       Fl_Group {} {
-        label {tab&2}
+        label {tabbbbb&2}
         callback {if (Fl::callback_reason()==FL_REASON_CLOSED)
   o->parent()->remove(o);}
         tooltip {tab2 tests among other things the cooperation of modal windows and tabs} xywh {10 60 315 235} selection_color 2 when 16 hide
```

**Comment by Albrecht-S** [comment]
Note: you can see the effect better if the first character of tab 2 is for instance an 'O' or 'X', you increase the zoom factor to 200-300% and you shrink the window width slowly by dragging the right border. You can view the label's text moving to the left as you shrink the window size.

**Comment by wcout** [comment]
Somewhere along these changes a minor regression crept in: With scheme "plastic" the top of the tabs widget is overlapping the background:
![tabs_scheme_plastic_regr](https://github.com/fltk/fltk/assets/9895976/f03d72e3-058b-4f98-afaf-b83d4e04a0b4)

unittest schemes test, left: FLTK 1.3  (and FLTK 1.4 before), right: current FLTK 1.4

Presumably it was caused by this commit: https://github.com/fltk/fltk/commit/443ce623003a1628729ce4fc0e717afd78483990

**Comment by MatthiasWM** [comment]
Thanks for finding this. I did. change to code for background drawing to be more stable and not relying on what may have been drawn previously at the cost of not drawing background images. I am busy until the 26th, but I have an idea how I will fix it then.

**Comment by Albrecht-S** [comment]
Note: I reduced the priority to 2 (low) since only minor ("cosmetic") issues are left. I also removed "fixed" and "waiting for confirmation" labels.

Summary: There are two remaining issues:
- [ ] Artifacts in tab label drawing (https://github.com/fltk/fltk/issues/718#issuecomment-1704207151)
- [ ] Background drawing in tab label area (https://github.com/fltk/fltk/issues/718#issuecomment-1722424905)

**Comment by MatthiasWM** [comment]
The first effect would actually be visible in real life tabs as well if the first tab is shorter than the second tab below. I will see I there is a constant way to fix that.

The second issue is the result of solving a compromise the "other" way, but I have and idea how to solve that with perfect drawing, but then it will cost more CPU time. There is always one disadvantage ;-)

**Comment by MatthiasWM** [comment]
Both issues seem to solved nicely. I tried with some more exotic box types like FL_ROUNDED_BOX, and amazingly, it still looks surprisingly well. Tested with all schemes in test/unittests and test/tabs.

**Comment by Albrecht-S** [comment]
I confirm the issue is fixed.

**Comment by wcout** [comment]
Really sorry to bother again... but I just tested overflow mode "drag" with the latest changes:
If adding a few tabs and then trying to drag to the left **all** tabs disappear!

**Comment by MatthiasWM** [comment]
Thanks for testing that. i forgot to add the drag Offsete in the clipping code. Doh. I'll fix it ASAP.

**Comment by MatthiasWM** [comment]
Fixed in [76b16d8](https://github.com/fltk/fltk/commit/76b16d8cef24c4c548e21dad90cf2f8731ffceb9)
